### PR TITLE
Major gates.rst revision

### PR DESCRIPTION
**Pull request by levbishop**

### CODE REVIEW

File: source/language/gates.rst [hunk 1]
```diff
@@ -5,172 +5,103 @@
...
-quantum register and ``b`` is a qubit, the statement means apply ``CX a[j], b;`` for each
-index ``j`` into register ``a``.
+Parameterized gates
+~~~~~~~~~~~~~~~~~~~
 
-.. _fig_cnot-dist:
-.. multifigure::
-   :rowitems: 2
+As well as gates that each represent a fixed unitary, OpenQASM also supports gates that represent *families* of unitaries, parameterized
+by angle variables.  Todistinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are
```
Comment (mhillenbrand) [inline source/language/gates.rst:56]: minor typo
```suggestion
by angle variables. To distinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are
```

File: source/language/gates.rst [hunk 2]
```diff
@@ -5,172 +5,103 @@
...
 
-Hierarchically defined unitary gates
-------------------------------------
+The definition does not imply that ``h`` is
+implemented by an instruction ``U(π/2, 0, π)`` on the quantum computer. The implementation is up to
+the user and/or compiler, given information about the instructions supported by a particular target.
+A minimal compiler implementation might simply expand ``gate`` definitions repeatedly until reaching
+definitions for which :ref:`defcal blocks <pulse-gates>` are known. A more sophisticated implementation
+might use the `gate` definitions of the gates with associated ``defcal`` blocks to
+to build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.
```
Comment (mhillenbrand) [inline source/language/gates.rst:100]: nit: superfluous `to`
```suggestion
build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.
```

File: source/language/gates.rst [hunk 3]
```diff
@@ -5,172 +5,103 @@
...
+In case that the expanded set of gates is not mutually-commuting, then its important to note that the broadcasting is defined
+in increasing qubit index order ``qr[0]``, ``qr[1]``, ....
 
-   \left(\begin{array}{cccc}
-   1 & 0 & 0 & 0 \\
-   0 & 0 & 0 & 1 \\
-   0 & 0 & 1 & 0 \\
-   0 & 1 & 0 & 0 \end{array}\right)
+//FIXME: Is that what we really want? Maybe we could use attributes to mark which arguments are safe (commuting) to broadcast over
+and error in the case of non-commuting broadcasting?
```
Comment (blakejohnson) [inline source/language/gates.rst:50]: I think this is what we agreed to, at least.

File: source/language/gates.rst [hunk 4]
```diff
@@ -202,15 +133,19 @@ In general, new gates are defined by statements of the form
...
      body
    }
 
 where the optional parameter list ``params`` is a comma-separated list of variable
 parameters, and the argument list ``qargs`` is a comma-separated list of qubit
-arguments. The parameters are identifiers with arbitrary-precision numeric types.
+arguments. The parameters are identifiers that behave as ``angle`` type with unknown
+size. A compiler might recognize certain constructs and replace them with mathematically-
+equivalent versions that would be true for arbitrary precision, or it might do calculations
+at a fixed angle size, for example corresponding to the size of angle parameters to the corresponding
```
Comment (blakejohnson) [inline source/language/gates.rst:146]:
```suggestion
at a fixed ``angle`` size, for example corresponding to the size of angle parameters to the corresponding
```

File: source/language/gates.rst [hunk 5]
```diff
@@ -240,58 +175,32 @@ and these symbols are scoped only to the subroutine body.
...
++++++++++++++++++
+
 The modifier ``ctrl @`` replaces its gate argument :math:`U` by a
 controlled-:math:`U` gate. If the control bit is 0, nothing happens to the target bit.
 If the control bit is 1, :math:`U` acts on the target bit. Mathematically, the controlled-:math:`U`
 gate is defined as :math:`C_U = I \otimes U^c`, where :math:`c` is the integer value of the control
 bit and :math:`C_U` is the controlled-:math:`U` gate. The new quantum argument is prepended to the
 argument list for the controlled-:math:`U` gate. The quantum argument can be a register, and in this
-case controlled gate broadcast over it (as it shown in examples above for CNOT gate). The modified
+case controlled gate broadcast over it (as for all gates //FIXME). The modified
```
Comment (blakejohnson) [inline source/language/gates.rst:199]: What is the plan here?

File: source/language/gates.rst [hunk 6]
```diff
@@ -378,7 +293,73 @@ repetitions of ``inv @ U`` for :math:`k < 0`.
...
+   }
+
+defines ``CX`` as the standard CNOT gate.
+
+Relation of the built-in gates to hardware-native gates
+----------------------------
+
+For *non-parameterized gates*, the choice of ``U`` and ``gphase`` as the built-in gates, along with one
+two-qubit entangling gate CNOT as defined gives a universal gate set that can represent general n-qubit
+unitary with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation
```
Comment (blakejohnson) [inline source/language/gates.rst:342]:
```suggestion
unitaries with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation
```

File: source/language/gates.rst (hunk 5 above)
Comment (levbishop) [reply source/language/gates.rst]: The plan is that all gates broadcast with the same semantics

**Comment by levbishop** [comment]
I think I've addressed all remaining issues.
Possible controversial points:
1) I dropped the requirement that user-defined gates must have at least one qubit. It feels more consistent to treat gphase as a 0-qubit gate, and to allow user-defined 0-qubit gates eg: `gate negate {gphase(pi);}`
2) I kept the global phase of the standard gates. This maintains the backwards compatibility, but at the cost of having a few gates which are non-2pi-periodic: rx(), ry(), rz() (and by extension crx(), cry(), crz()), cu(), u2(), u3(). If we are already sucking up the breaking change of U() global phase maybe we should just bite the bullet and do the same for these gates too?

File: source/language/gates.rst [hunk 7]
```diff
@@ -202,18 +135,21 @@ In general, new gates are defined by statements of the form
...
      body
    }
 
 where the optional parameter list ``params`` is a comma-separated list of variable
 parameters, and the argument list ``qargs`` is a comma-separated list of qubit
-arguments. The parameters are identifiers with arbitrary-precision numeric types.
+arguments. The parameters are identifiers that behave as ``angle`` type with unknown
+size. A compiler might recognize certain constructs and replace them with mathematically-
+equivalent versions that would be true for arbitrary precision, or it might do calculations
+at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters to the corresponding
```
Comment (blakejohnson) [inline source/language/gates.rst:148]:
```suggestion
at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters in the corresponding
```

File: source/language/gates.rst [hunk 8]
```diff
@@ -378,7 +294,87 @@ repetitions of ``inv @ U`` for :math:`k < 0`.
...
+For *parameterized gates*, the choice of built-in gates *does* constrain which hardware-native gates are well-
+supported, because conversion between parameterized basis sets in general can be involved, requiring careful
+selection of branch cuts and other logic that would not likely be feasible to specify as compact mathematical
+expressions, nor to evaluate at runtime for cases where the parameters depend on quantum measurements.
+
+For many current platforms the qubits are defined relative to a
+rotating frame and the rotating wave approximation (RWA) holds. This is the domain covered by the OpenPulse
+specification. For this case, the only supported form of run-time parameterization
+will likely be via a ``rz(ϕ)`` implemented by specialized frame-tracking hardware.
+This gate is covered by the built-in ``U`` as a special case ``U(0, ϕ, θ)``
```
Comment (blakejohnson) [inline source/language/gates.rst:373]: Do you mean to have zeros for the first two arguments of `U`?

**Review by blakejohnson — COMMENTED** [review]
I've double checked all the expressions in `stdgates.inc` and it looks correct to me. A have just a couple small questions I would like resolved.

File: source/language/gates.rst (hunk 8 above)
Comment (levbishop) [reply source/language/gates.rst:373]: I didn't but agree that would be clearer

Comment (levbishop) [inline source/language/gates.rst:373]:
```suggestion
This gate is covered by the built-in ``U`` as a special case ``U(0, 0, ϕ)``
```

**Review by blakejohnson — APPROVED** [review]

**Comment by blakejohnson** [comment]
Thanks, Lev. I'm glad to get this merged!

**Pull request merged**